tests/end_to_end/end_to_end.cpp

```diff
@@ -1064,4 +1064,172 @@ TEST_CASE("link_tests", "[end_to_end]")
+    REQUIRE(port_map_value == 200);
+
+    Platform::_close(outer_map_fd);
+    Platform::_close(inner_map_fd);
+    Platform::_close(port_map_fd);
+
+    REQUIRE(ebpf_object_unpin("/ebpf/global/outer_map") == EBPF_SUCCESS);
+    REQUIRE(ebpf_object_unpin("/ebpf/global/port_map") == EBPF_SUCCESS);
+}
+
+TEST_CASE("auto_pinned_maps_invalid", "[end_to_end]")
+{
+    _test_helper_end_to_end test_helper;
+    single_instance_hook_t hook(EBPF_PROGRAM_TYPE_XDP, EBPF_ATTACH_TYPE_XDP);
+    program_info_provider_t xdp_program_info(EBPF_PROGRAM_TYPE_XDP);
+
+    // First create and pin the maps manually.
+    int outer_map_fd = bpf_create_map(BPF_MAP_TYPE_HASH_OF_MAPS, sizeof(__u32), sizeof(__u32), 1, 0);
+    REQUIRE(outer_map_fd > 0);
+
+    int inner_map_fd = bpf_create_map(BPF_MAP_TYPE_ARRAY, sizeof(__u32), sizeof(__u32), 1, 0);
+    REQUIRE(inner_map_fd > 0);
+
+    // Pin the outer map.
+    // We have not yet inserted any inner map in the outer map.
+    int error = bpf_obj_pin(outer_map_fd, "/ebpf/global/outer_map");
+    REQUIRE(error == 0);
+
+    int port_map_fd = bpf_create_map(BPF_MAP_TYPE_ARRAY, sizeof(__u32), sizeof(__u32), 1, 0);
+    REQUIRE(outer_map_fd > 0);
+
+    // Pin port map
+    error = bpf_obj_pin(port_map_fd, "/ebpf/global/port_map");
+    REQUIRE(error == 0);
+
+    // Load BPF object from ELF file. Loading the program should fail as the
+    // outer map still does not have any inner map template.
+    struct bpf_object* object = nullptr;
+    fd_t program_fd;
+    const char* log_buffer = nullptr;
+    ebpf_result_t result = ebpf_program_load(
+        SAMPLE_PATH "map_reuse.o",
+        &EBPF_PROGRAM_TYPE_XDP,
+        nullptr,
+        EBPF_EXECUTION_ANY,
+        &object,
+        &program_fd,
+        &log_buffer);
+
+    ebpf_free_string(log_buffer);
+    REQUIRE(result == EBPF_INVALID_ARGUMENT);
+
+    Platform::_close(outer_map_fd);
+    Platform::_close(inner_map_fd);
+    Platform::_close(port_map_fd);
+
+    REQUIRE(ebpf_object_unpin("/ebpf/global/outer_map") == EBPF_SUCCESS);
+    REQUIRE(ebpf_object_unpin("/ebpf/global/port_map") == EBPF_SUCCESS);
+}
+
+TEST_CASE("auto_pinned_maps_invalid_2", "[end_to_end]")
+{
+    _test_helper_end_to_end test_helper;
+    single_instance_hook_t hook(EBPF_PROGRAM_TYPE_XDP, EBPF_ATTACH_TYPE_XDP);
+    program_info_provider_t xdp_program_info(EBPF_PROGRAM_TYPE_XDP);
+
+    // First create and pin the maps manually.
+    int outer_map_fd = bpf_create_map(BPF_MAP_TYPE_HASH_OF_MAPS, sizeof(__u32), sizeof(__u32), 1, 0);
+    REQUIRE(outer_map_fd > 0);
+
+    int inner_map_fd = bpf_create_map(BPF_MAP_TYPE_ARRAY, sizeof(__u32), sizeof(__u32), 1, 0);
+    REQUIRE(inner_map_fd > 0);
+
+    // Verify we can insert the inner map into the outer map.
+    __u32 outer_key = 0;
+    int error = bpf_map_update_elem(outer_map_fd, &outer_key, &inner_map_fd, 0);
+    REQUIRE(error == 0);
+
+    // Pin the outer map
```

Comment:
```suggestion
    // Pin the outer map.
```
nit: Some comments end in period and others don't.  Go through and make all the new ones consistent.